blockchain/v0/reactor.go

```diff
@@ -49,339 +72,446 @@ func (e peerError) Error() string {
-	msgBytes, err := bc.EncodeMsg(&bcproto.StatusResponse{
-		Base:   bcR.store.Base(),
-		Height: bcR.store.Height()})
-	if err != nil {
-		bcR.Logger.Error("could not convert msg to protobuf", "err", err)
-		return
-	}
+// respondToPeer loads a block and sends it to the requesting peer, if we have it.
+// Otherwise, we'll respond saying we do not have it.
+func (r *Reactor) respondToPeer(msg *bcproto.BlockRequest, peerID p2p.NodeID) {
+	block := r.store.LoadBlock(msg.Height)
+	if block != nil {
+		blockProto, err := block.ToProto()
+		if err != nil {
+			r.Logger.Error("failed to convert msg to protobuf", "err", err)
+			return
+		}
 
-	_ = peer.Send(BlockchainChannel, msgBytes)
-	// it's OK if send fails. will try later in poolRoutine
+		r.blockchainCh.Out() <- p2p.Envelope{
+			To:      peerID,
+			Message: &bcproto.BlockResponse{Block: blockProto},
+		}
 
-	// peer is added to the pool once we receive the first
-	// bcStatusResponseMessage from the peer and call pool.SetPeerRange
-}
+		return
+	}
 
-// RemovePeer implements Reactor by removing peer from the pool.
-func (bcR *BlockchainReactor) RemovePeer(peer p2p.Peer, reason interface{}) {
-	bcR.pool.RemovePeer(peer.ID())
+	r.Logger.Info("peer requesting a block we do not have", "peer", peerID, "height", msg.Height)
+	r.blockchainCh.Out() <- p2p.Envelope{
+		To:      peerID,
+		Message: &bcproto.NoBlockResponse{Height: msg.Height},
+	}
 }
 
-// respondToPeer loads a block and sends it to the requesting peer,
-// if we have it. Otherwise, we'll respond saying we don't have it.
-func (bcR *BlockchainReactor) respondToPeer(msg *bcproto.BlockRequest,
-	src p2p.Peer) (queued bool) {
+// handleBlockchainMessage handles enevelopes sent from peers on the
+// BlockchainChannel. It returns an error only if the Envelope.Message is unknown
+// for this channel. This should never be called outside of handleMessage.
+func (r *Reactor) handleBlockchainMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *bcproto.BlockRequest:
+		r.respondToPeer(msg, envelope.From)
 
-	block := bcR.store.LoadBlock(msg.Height)
-	if block != nil {
-		bl, err := block.ToProto()
+	case *bcproto.BlockResponse:
+		block, err := types.BlockFromProto(msg.Block)
 		if err != nil {
-			bcR.Logger.Error("could not convert msg to protobuf", "err", err)
-			return false
+			r.Logger.Error("failed to convert block from proto", "err", err)
+			return err
 		}
 
-		msgBytes, err := bc.EncodeMsg(&bcproto.BlockResponse{Block: bl})
+		protoMsg := new(bcproto.Message)
+		if err := protoMsg.Wrap(msg); err != nil {
+			r.Logger.Error("failed to wrap proto message", "err", err)
+			return nil
+		}
+
+		bz, err := proto.Marshal(protoMsg)
 		if err != nil {
-			bcR.Logger.Error("could not marshal msg", "err", err)
-			return false
+			r.Logger.Error("failed to proto encode message", "err", err)
+			return nil
+		}
+
+		r.pool.AddBlock(envelope.From, block, len(bz))
```

Comment:
> In an case is it possible to use `msg.Block.Size()` instead of `len(bz)`

I don't think so. The [original/current code](https://github.com/tendermint/tendermint/blob/master/blockchain/v0/reactor.go#L240) passes the length in bytes of the _entire_ proto message.

Reply:
Yeah that's true. There would be a slight discrepancy in size. I was thinking however that such discrepancy could be tolerated if this was just used to monitor the byte rate of the peer. But if we're unsure then probably best to leave it as is